File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def sameTree(a, b):
            if not a and not b:
                return True

            if not a or not b:
                return False

            if a.val != b.val:
                return False

            return (
                sameTree(a.left, b.left)
                and sameTree(a.right, b.right)
            )

        if not subRoot:
            return True

        if not root:
            return False

        if sameTree(root, subRoot):
            return True

        return (
            self.isSubtree(root.left, subRoot)
            or self.isSubtree(root.right, subRoot)
        )

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-1.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func buildTree(preorder []int, inorder []int) *TreeNode {
    m := make(map[int]int)
    for i := range inorder {
        m[inorder[i]] = i
    }

    var helper func(preL, preR, inL, inR int) *TreeNode
    helper = func(preL, preR, inL, inR int) *TreeNode {
        if preL > preR {
            return nil
        }

        rootV := preorder[preL]
        node := &TreeNode{ Val:rootV }

        rootI := m[rootV]
        lSize := rootI - inL

        node.Left = helper(
            preL + 1,
            preL + lSize,
            inL,
            rootI - 1,
        )

        node.Right = helper(
            preL + lSize + 1,
            preR,
            rootI + 1,
            inR,
        )
        return node
    }
    return helper(0, len(preorder)-1, 0, len(inorder)-1)
}
